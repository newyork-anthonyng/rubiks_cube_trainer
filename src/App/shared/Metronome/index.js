import React, { Component } from "react";
import * as workerTimers from "worker-timers";
import Tone from "../Tone";

class Metronome extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isPlaying: false,
      isActive: false,
      bpm: 60
    };

    this.iosAudioContextUnlocked = false;
    this.audioContext = null;
    if (typeof AudioContext !== "undefined") {
      this.audioContext = new AudioContext();
    }
  }

  componentWillUnmount() {
    workerTimers.clearInterval(this.interval);
  }

  handleClick = () => {
    if (!this.iosAudioContextUnlocked) {
      const buffer = this.audioContext.createBuffer(1, 1, 22050);
      const node = this.audioContext.createBufferSource();
      node.buffer = buffer;
      node.start(0);
      this.iosAudioContextUnlocked = true;
    }

    this.setState(
      prevState => ({
        isActive: !prevState.isActive
      }),
      () => {
        if (this.state.isActive) {
          // start playing the timer
          this.interval = workerTimers.setInterval(() => {
            this.setState(prevState => ({ isPlaying: !prevState.isPlaying }));
          }, this.calculateTimeInterval());
        } else {
          // stop playing the timer
          workerTimers.clearInterval(this.interval);
          this.interval = null;
        }
      }
    );
  };

  calculateTimeInterval = () => {
    const time = 60 / this.state.bpm * 1000;
    // divide by 2 so that Tone is set to play at the beginning of each interval
    return time / 2;
  };

  handleBpmChange = e => {
    this.setState({ bpm: e.target.value }, () => {
      if (this.state.isActive) {
        workerTimers.clearInterval(this.interval);
        this.interval = workerTimers.setInterval(() => {
          this.setState(prevState => ({ isPlaying: !prevState.isPlaying }));
        }, this.calculateTimeInterval());
      }
    });
  };

  render() {
    const { isActive, bpm, isPlaying, length } = this.state;

    return (
      <div>
        <button onClick={this.handleClick}>
          {isActive ? "Pause" : "Play"}
        </button>
        <input
          type="range"
          value={bpm}
          min={45}
          max={160}
          onChange={this.handleBpmChange}
        />
        <span>BPM: {bpm}</span>

        <Tone
          play={isPlaying}
          length={length}
          audioContext={this.audioContext}
        />
      </div>
    );
  }
}

export default Metronome;
