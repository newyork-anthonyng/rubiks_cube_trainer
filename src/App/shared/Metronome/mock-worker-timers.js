// Mock "worker-timers" for SSR
export default {};
