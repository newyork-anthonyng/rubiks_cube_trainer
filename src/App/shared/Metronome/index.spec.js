jest.mock("worker-timers", () => {
  return {
    clearInterval: jest.fn(),
    setInterval: jest.fn()
  };
});
import * as workerTimers from "worker-timers";
import React from "react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import Metronome from "./";

beforeEach(() => {
  workerTimers.clearInterval.mockClear();
  workerTimers.setInterval.mockClear();

  // AudioContext mocks
  createBuffer.mockClear();
  createBufferSource.mockClear();
  bufferStart.mockClear();
});

it("should render correctly when not playing", () => {
  const wrapper = shallow(<Metronome />);

  expect(toJSON(wrapper)).toMatchSnapshot();
});

it("should render correctly when playing", () => {
  const wrapper = shallow(<Metronome />);

  const buttonEl = wrapper.find("button");
  buttonEl.simulate("click");

  expect(toJSON(wrapper)).toMatchSnapshot();
});

it("should update BPM when changing input", () => {
  const wrapper = shallow(<Metronome />);

  const rangeInputEl = wrapper.find("input[type='range']");
  rangeInputEl.simulate("change", {
    target: { value: 42 }
  });

  expect(toJSON(wrapper)).toMatchSnapshot();
});

it("should update workerTimer if changing BPM while metronome is active", () => {
  const wrapper = shallow(<Metronome />);

  const buttonEl = wrapper.find("button");
  buttonEl.simulate("click");

  workerTimers.clearInterval.mockClear();
  workerTimers.setInterval.mockClear();

  const rangeInputEl = wrapper.find("input[type='range']");
  rangeInputEl.simulate("change", {
    target: { value: 42 }
  });

  expect(workerTimers.clearInterval).toHaveBeenCalledTimes(1);
  expect(workerTimers.setInterval).toHaveBeenCalledTimes(1);
  expect(workerTimers.setInterval.mock.calls[0]).toMatchSnapshot();
});

// iOS bug. See this issue https://github.com/cwilso/metronome/issues/14
it("should play an empty buffer when first clicking button", () => {
  const wrapper = shallow(<Metronome />);

  const buttonEl = wrapper.find("button");
  buttonEl.simulate("click");

  expect(createBuffer).toHaveBeenCalledTimes(1);
  expect(createBufferSource).toHaveBeenCalledTimes(1);
  expect(bufferStart).toHaveBeenCalledTimes(1);

  createBuffer.mockClear();
  createBufferSource.mockClear();
  bufferStart.mockClear();

  buttonEl.simulate("click");

  expect(createBuffer).not.toHaveBeenCalled();
  expect(createBufferSource).not.toHaveBeenCalled();
  expect(bufferStart).not.toHaveBeenCalled();
});

it("should update workerTimer correctly when clicking on button to pause metronome", () => {
  const wrapper = shallow(<Metronome />);

  // play Metronome
  const buttonEl = wrapper.find("button");
  buttonEl.simulate("click");
  workerTimers.clearInterval.mockClear();
  workerTimers.setInterval.mockClear();

  // pause Metronome
  buttonEl.simulate("click");

  expect(workerTimers.clearInterval).toHaveBeenCalledTimes(1);
  expect(workerTimers.setInterval).not.toHaveBeenCalled();
});

it("should update workerTimer correctly when clicking on button to play metronome", () => {
  const wrapper = shallow(<Metronome />);

  // play Metronome
  const buttonEl = wrapper.find("button");
  buttonEl.simulate("click");

  expect(workerTimers.setInterval).toHaveBeenCalledTimes(1);
  expect(workerTimers.setInterval.mock.calls[0]).toMatchSnapshot();
  expect(workerTimers.clearInterval).not.toHaveBeenCalled();

  // setInterval function should toggle isPlaying state
  const setIntervalCb = workerTimers.setInterval.mock.calls[0][0];
  setIntervalCb();
  wrapper.update();
  expect(toJSON(wrapper)).toMatchSnapshot();

  setIntervalCb();
  wrapper.update();
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it("should clear interval when unmounting", () => {
  const wrapper = mount(<Metronome />);

  wrapper.unmount();

  expect(workerTimers.clearInterval).toHaveBeenCalledTimes(1);
});
