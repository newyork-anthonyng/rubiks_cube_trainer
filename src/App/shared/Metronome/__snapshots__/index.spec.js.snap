// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly when not playing 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Play
  </button>
  <label>
    Update BPM: 
    60
    <input
      max={160}
      min={45}
      onChange={[Function]}
      type="range"
      value={60}
    />
  </label>
  <Tone
    audioContext={
      Object {
        "createBuffer": [Function],
        "createBufferSource": [Function],
      }
    }
    frequency={400}
    length={0.05}
    onStop={[Function]}
    play={false}
  />
</div>
`;

exports[`should render correctly when playing 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Pause
  </button>
  <label>
    Update BPM: 
    60
    <input
      max={160}
      min={45}
      onChange={[Function]}
      type="range"
      value={60}
    />
  </label>
  <Tone
    audioContext={
      Object {
        "createBuffer": [Function],
        "createBufferSource": [Function],
      }
    }
    frequency={400}
    length={0.05}
    onStop={[Function]}
    play={false}
  />
</div>
`;

exports[`should update BPM when changing input 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Play
  </button>
  <label>
    Update BPM: 
    42
    <input
      max={160}
      min={45}
      onChange={[Function]}
      type="range"
      value={42}
    />
  </label>
  <Tone
    audioContext={
      Object {
        "createBuffer": [Function],
        "createBufferSource": [Function],
      }
    }
    frequency={400}
    length={0.05}
    onStop={[Function]}
    play={false}
  />
</div>
`;

exports[`should update workerTimer correctly when clicking on button to play metronome 1`] = `
Array [
  [Function],
  500,
]
`;

exports[`should update workerTimer correctly when clicking on button to play metronome 2`] = `
<div>
  <button
    onClick={[Function]}
  >
    Pause
  </button>
  <label>
    Update BPM: 
    60
    <input
      max={160}
      min={45}
      onChange={[Function]}
      type="range"
      value={60}
    />
  </label>
  <Tone
    audioContext={
      Object {
        "createBuffer": [Function],
        "createBufferSource": [Function],
      }
    }
    frequency={400}
    length={0.05}
    onStop={[Function]}
    play={true}
  />
</div>
`;

exports[`should update workerTimer correctly when clicking on button to play metronome 3`] = `
<div>
  <button
    onClick={[Function]}
  >
    Pause
  </button>
  <label>
    Update BPM: 
    60
    <input
      max={160}
      min={45}
      onChange={[Function]}
      type="range"
      value={60}
    />
  </label>
  <Tone
    audioContext={
      Object {
        "createBuffer": [Function],
        "createBufferSource": [Function],
      }
    }
    frequency={400}
    length={0.05}
    onStop={[Function]}
    play={false}
  />
</div>
`;

exports[`should update workerTimer if changing BPM while metronome is active 1`] = `
Array [
  [Function],
  714.2857142857143,
]
`;
